<?xml version="1.0" ?>
<launch>
  <!-- LAUNCH INTERFACE -->

  <!-- The robot name is required for gazebo to spawn a model. -->
  <arg name="rightRobotName" default="right_lwr"/>
  <arg name="leftRobotName" default="left_lwr"/>
  <arg name="rightRobotOrigin.x" default="0.5"/>
  <arg name="rightRobotOrigin.y" default="0"/>
  <arg name="rightRobotOrigin.z" default="0"/>
  <arg name="rightRobotOrigin.ax" default="0"/>
  <arg name="rightRobotOrigin.ay" default="0"/>
  <arg name="rightRobotOrigin.az" default="0"/>
  <arg name="leftRobotOrigin.x" default="-0.5"/>
  <arg name="leftRobotOrigin.y" default="0.5"/>
  <arg name="leftRobotOrigin.z" default="0"/>
  <arg name="leftRobotOrigin.ax" default="0"/>
  <arg name="leftRobotOrigin.ay" default="0"/>
  <arg name="leftRobotOrigin.az" default="${M_PI}"/>
  <arg name="robotsMiddle.x" default="$(eval 0.5*arg('rightRobotOrigin.x')+0.5*arg('leftRobotOrigin.x'))"/>
  <arg name="robotsMiddle.y" default="$(eval 0.5*arg('rightRobotOrigin.y')+0.5*arg('leftRobotOrigin.y'))"/>
  <arg name="robotsMiddle.z" default="$(eval 0.5*arg('rightRobotOrigin.z')+0.5*arg('leftRobotOrigin.z'))"/>
  <arg name="rightTrocarOffset.x" default="0"/>
  <arg name="rightTrocarOffset.y" default="0"/>
  <arg name="rightTrocarOffset.z" default="0.2"/>
  <arg name="leftTrocarOffset.x" default="-0.1"/>
  <arg name="leftTrocarOffset.y" default="0.3"/>
  <arg name="leftTrocarOffset.z" default="0.3"/>

  <!-- The recommended way of specifying the URDF file is to pass it in
       directly. -->
  <arg name="robotPath" default="$(find surgical_simulator)/models"/>
  <arg name="lwrRobotUrdfFile" default="$(arg robotPath)/kuka_lwr_on_table_with_tool.urdf.xacro" />
  <arg name="environmentUrdfFile" default="$(arg robotPath)/surgical_environment.urdf.xacro" />
  <arg name="pancreasUrdfFile" default="$(arg robotPath)/h_lower_limb.urdf.xacro" />

  <arg name="controllers" default="joint_controllers"/>
  <arg name="t1_limits" default="false"/>

  <arg name="rviz_config" default="$(find lwr_launch)/launch/rviz/rviz_config.rviz"/>
  <arg name="rviz_bringup" default="true"/> <!--If false, do not launch rviz-->

  <arg name="right_hw_interface_file" default="$(find surgical_simulator)/config/right_hw_interface.yaml"/>
  <arg name="right_controller_config_file" default="$(find surgical_simulator)/config/right_controllers.yaml"/>
  <arg name="left_hw_interface_file" default="$(find surgical_simulator)/config/left_hw_interface.yaml"/>
  <arg name="left_controller_config_file" default="$(find surgical_simulator)/config/left_controllers.yaml"/>
  <arg name="gazebo_world_file" default="$(find surgical_simulator)/worlds/surgical_simulator_world.world"/>
  <arg name="t1_limits_file" default="$(find surgical_simulator)/config/t1_joint_limits.yaml"/>
  <arg name="GAZEBO_GUI" default="false"/>

  <!-- Load updated joint limits (override information from single_lwr_moveit) to respect T1 mode limits -->
  <group if="$(arg t1_limits)" ns="robot_description_planning">
    <rosparam command="load" file="$(arg t1_limits_file)"/>
  </group>

  <!-- load robotHW configurations to rosparam server -->
  <rosparam command="load" file="$(arg right_hw_interface_file)"/>
  <rosparam command="load" file="$(arg left_hw_interface_file)"/>

  <!-- load all controller configurations to rosparam server -->
  <rosparam command="load" file="$(arg right_controller_config_file)"/>
  <rosparam command="load" file="$(arg left_controller_config_file)"/>

    <!-- <node name="worldRightTransform" pkg="foot_control_xy" type="worldActorTransform" args="foot -x 2.0 -y 0.0 -z 0 -qx 0.0 -qy 0.0 -qz 0.0 -qw 1.0"/> -->

  <group ns="environment">
    <param name="robot_description" command="$(find xacro)/xacro.py $(arg environmentUrdfFile) origin.x:=0 origin.y:=0 origin.z:=0"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <node name="environment_spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-param /environment/robot_description -urdf -model environment" respawn="false" output="screen" />
  </group>
  
  <!-- ROBOT -->
  <group ns="right_lwr">
    <param name="robot_description" command="$(find xacro)/xacro.py $(arg lwrRobotUrdfFile) robotName:=$(arg rightRobotName) origin.x:=$(arg rightRobotOrigin.x) origin.y:=$(arg rightRobotOrigin.y) origin.z:=$(arg rightRobotOrigin.z) origin.ax:=$(arg rightRobotOrigin.ax) origin.ay:=$(arg rightRobotOrigin.ay) origin.az:=$(arg rightRobotOrigin.az) addCamera:=0 addFrame:=1"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <node name="controller_spawner_js" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="--namespace=/right_lwr kuka_joint_state_controller"/>
    <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="--namespace=/right_lwr $(arg controllers)"/>
    <node name="right_spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-param /right_lwr/robot_description -urdf -model $(arg rightRobotName)" respawn="false" output="screen" />
    <node name="publish_trocar_frame" pkg="surgical_simulator" type="publish_trocar_frame" args="$(arg rightRobotName) -x $(arg rightTrocarOffset.x) -y $(arg rightTrocarOffset.y) -z $(arg rightTrocarOffset.z)"/>

  </group>

  <group ns="left_lwr">
    <param name="robot_description" command="$(find xacro)/xacro.py $(arg lwrRobotUrdfFile) robotName:=$(arg leftRobotName) origin.x:=$(arg leftRobotOrigin.x) origin.y:=$(arg leftRobotOrigin.y) origin.z:=$(arg leftRobotOrigin.z) origin.ax:=$(arg leftRobotOrigin.ax) origin.ay:=$(arg leftRobotOrigin.ay) origin.az:=$(arg leftRobotOrigin.az) addCamera:=1"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <node name="controller_spawner_js" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="kuka_joint_state_controller"/>
    <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="$(arg controllers)"/>
    <node name="left_spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-param /left_lwr/robot_description -urdf -model $(arg leftRobotName)" respawn="false" output="screen" />
    <node name="publish_trocar_frame" pkg="surgical_simulator" type="publish_trocar_frame" args="$(arg leftRobotName) -x $(arg leftTrocarOffset.x) -y $(arg leftTrocarOffset.y) -z $(arg leftTrocarOffset.z)"/>
  </group>

  <!-- GAZEBO -->
  <!-- enable/disable gui at will, the rviz listens to the simulation -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(arg gazebo_world_file)"/>
    <arg name="paused" value="false"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="$(arg GAZEBO_GUI)"/>
    <arg name="headless" value="false"/>
    <arg name="debug" value="false"/>
  </include>

  <!-- <node pkg="tf" type="static_transform_publisher" name="right_to_left_transform" args="$(arg leftRobotOrigin.x) $(arg leftRobotOrigin.y) $(arg leftRobotOrigin.z) 0 0 0 /world /left_robot_basis 200"/>  -->


<!--   <group ns="pancreas">
    <param name="robot_description" command="$(find xacro)/xacro.py $(arg pancreasUrdfFile)"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <node name="pancreas_spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-x 3 -y 3 -z 3 -param /pancreas/robot_description -urdf -model pancreas" respawn="false" output="screen" />
  </group> -->

  <!-- LAUNCH RVIZ -->

  <node name="lwr_rviz" pkg="rviz" type="rviz" respawn="false" args="-d $(arg rviz_config)"
        output="screen"
        if="$(arg rviz_bringup)"/>

    <!-- <node name="righTrocarPublish" pkg="surgical_simulator" type="publish_trocar_frame" args="right_trocar -x 0 -y 0 -z 0"/> -->

</launch>
